### PR TITLE
Issue 13379: Metrics: noise stacktrace if Prometheus closes the connection due to a timeout

### LINKED ISSUE #13379: Metrics: noise stacktrace if Prometheus closes the connection due to a timeout
**Describe the bug**
if Prometheus closes the connection due to a timeout or some other reason you will see these errors in the logs and this is very noisy.
We can simply report a better message in the logs.

10:35:18.642 [prometheus-stats-58-1] ERROR org.apache.pulsar.broker.stats.prometheus.PrometheusMetricsServlet - Failed to generate prometheus stats
org.eclipse.jetty.io.EofException: null
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:279) ~[jetty-io-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:422) ~[jetty-io-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:277) ~[jetty-io-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:381) ~[jetty-io-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpConnection$SendCallback.process(HttpConnection.java:826) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241) ~[jetty-util-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:223) ~[jetty-util-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpConnection.send(HttpConnection.java:550) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpChannel.sendResponse(HttpChannel.java:915) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpChannel.write(HttpChannel.java:987) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpOutput.channelWrite(HttpOutput.java:285) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpOutput.channelWrite(HttpOutput.java:269) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:897) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.apache.pulsar.broker.stats.prometheus.PrometheusMetricsGenerator.generate(PrometheusMetricsGenerator.java:127) .... 4.1.60.Final.jar:4.1.60.Final]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.io.IOException: Broken pipe
	at sun.nio.ch.FileDispatcherImpl.writev0(Native Method) ~[?:?]
	at sun.nio.ch.SocketDispatcher.writev(SocketDispatcher.java:51) ~[?:?]
	at sun.nio.ch.IOUtil.write(IOUtil.java:182) ~[?:?]
	at sun.nio.ch.IOUtil.write(IOUtil.java:130) ~[?:?]
	at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:493) ~[?:?]
	at java.nio.channels.SocketChannel.write(SocketChannel.java:507) ~[?:?]
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:273) ~[jetty-io-9.4.43.v20210629.jar:9.4.43.v20210629]
	... 23 more

**Expected behavior**
The logs should not be full of this spammy log

**Additional context**
The error may occour if you have many functions in the broker and collecting the metrics is too slow.

### PR DESCRIPTION
**Pull request by eolivelli**
Fixes #13379 

### Motivation

When there is a timeout in serving the metrics Pulsar will print a long stacktrace that is scary and meaningless to the users.

### Modifications

1. Provide a better log message: no stack trace, include the processing time
2. ensure that context.complete() is always executed, in a finally block
3. Handle gracefully the case of "AsyncContext completed and/or Request lifecycle recycled" in case of  server side timeout (metricsServletTimeoutMs)

### Verifying this change
This change is a trivial rework / code cleanup without any test coverage.

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java [hunk 1]
```diff
@@ -70,18 +71,27 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
...
                 res.setContentType("text/plain");
                 PrometheusMetricsGenerator.generate(pulsar, shouldExportTopicMetrics, shouldExportConsumerMetrics,
                         shouldExportProducerMetrics, splitTopicAndPartitionLabel, res.getOutputStream(),
                         metricsProviders);
-                context.complete();
-
             } catch (Exception e) {
-                log.error("Failed to generate prometheus stats", e);
+                long end = System.currentTimeMillis();
+                long time = end - start;
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java:84]: I wonder if we should use Guava's Stopwatch for such use cases? Javadoc: https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/base/Stopwatch.html

**Review by lhotari — APPROVED** [review]

**Review by nicoloboschi — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java (hunk 1 above)
Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java:84]: I don't like very much Guava Stopwatch (in the past the broke compatibility), and in general I believe that we should keep off Guava as much as possible (still for the fact that they break the APIs quite often)

**Review by nicoloboschi — APPROVED** [review]
LGTM

**Review by congbobo184 — APPROVED** [review]

**Review by dlg99 — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java [hunk 2]
```diff
@@ -70,19 +71,38 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
...
             try {
                 res.setStatus(HttpStatus.OK_200);
                 res.setContentType("text/plain");
                 PrometheusMetricsGenerator.generate(pulsar, shouldExportTopicMetrics, shouldExportConsumerMetrics,
                         shouldExportProducerMetrics, splitTopicAndPartitionLabel, res.getOutputStream(),
                         metricsProviders);
-                context.complete();
-
             } catch (Exception e) {
-                log.error("Failed to generate prometheus stats", e);
```
[the remark this answers is not in the document]
Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java:83]: is it a really bad error and we must see this in production (and we also need the full stacktrace)

if we are not able to generate metrics (that should be a very light weight process) these is some action to take, if we leave DEBUG nobody will probably notice

**Pull request merged**